### PR TITLE
Configurable delete range

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
PTAL

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -309,6 +309,8 @@ pub struct ApplyDelegate {
     term: u64,
     pending_cmds: PendingCmdQueue,
     metrics: ApplyMetrics,
+
+    use_delete_range: bool,
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: Is it more suitable to put this to `ApplyContext`?

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -1974,7 +1996,7 @@ mod tests {
         let obs = ApplyObserver::default();
         host.registry
             .register_query_observer(1, Box::new(obs.clone()));
-        let mut apply_ctx = ApplyContext::new(&host);
+        let mut apply_ctx = ApplyContext::new(&host, true /* use delete range */);
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: can remove the comment

**Comment by siddontang** [comment]
Do we still need to support the test without DeleteRange now?

Rest LGTM

PTAL @huachaohuang @BusyJay

**Comment by huachaohuang** [comment]
LGTM

**Comment by siddontang** [comment]
@zhangjinpeng1987 I think we can merge this, and use delete range as default in our all tests now.

File: src/raftstore/store/worker/region.rs [hunk 3]
```diff
@@ -254,7 +260,9 @@ impl SnapContext {
             escape(&start_key),
             escape(&end_key)
         );
-        if let Err(e) = util::delete_all_in_range(&self.kv_db, &start_key, &end_key) {
+        if let Err(e) =
+            util::delete_all_in_range(&self.kv_db, &start_key, &end_key, self.use_delete_range)
```
Comment (BusyJay) [inline src/raftstore/store/worker/region.rs]: I don't think it's appropriate to use delete range here. For long running queries, delete all sst files directly may introduce wrong result.

Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/region.rs]: delete range will not break snapshot, so it is ok.

Comment (BusyJay) [reply src/raftstore/store/worker/region.rs]: Why? When SST files are deleted, keys in the bottom level can be seen by the snapshot.

Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/region.rs]: delete range never drop sst files directly.

Comment (zhangjinpeng1987) [reply src/raftstore/store/worker/region.rs]: Are you misunderstand `delete files in range` and `delete range`.

Comment (BusyJay) [reply src/raftstore/store/worker/region.rs]: Oh, I see. I thought `use_delete_range` is a hint for `delete_file_in_range`. It's OK to use `delete_range` here. Thanks for the explanation.

**Pull request merged**